### ISSUE TITLE
Usage of hard or symbolic linking leads to NuGet cache corruption

### ISSUE THREAD

**Issue by marcin-krystianc** [dotnet/msbuild#8273]
### Issue Description

MSBuild can use hard or symbolic links to avoid excessive file copies on disk. It is a good method to speed-up builds but at the same time, it is very easy to silently corrupt the NuGet cache. 
I have seen that this issue has been mentioned several times already, but it was never explained how and when exactly the NuGet cache gets corrupted:

Related issues:
- https://github.com/dotnet/msbuild/issues/1407
- https://github.com/dotnet/msbuild/issues/3788
- https://github.com/dotnet/msbuild/issues/7014
- https://github.com/NuGet/Home/issues/12047

### Steps to Reproduce

1. Build an executable project with hard or symbolic links enabled
2. Update the version of referenced NuGet package
3. Build the application again, but this time, without using hard/symbolic links.
MSBuild will try to update dependencies in the application's output folder. Sadly, instead of replacing existing links, it replaces actual files in NuGet cache (thus corrupting it).

- hard-links
```
dotnet nuget locals --clear all
dotnet new console
dotnet add package newtonsoft.json -v 13.0.1
dotnet build /p:CreateHardLinksForCopyLocalIfPossible=true
dotnet add package newtonsoft.json -v 13.0.2
dotnet build
```

- symbolic-links
```
dotnet nuget locals --clear all
dotnet new console
dotnet add package newtonsoft.json -v 13.0.1
dotnet build /p:CreateSymbolicLinksForCopyLocalIfPossible=true
dotnet add package newtonsoft.json -v 13.0.2
dotnet build
```

In both cases file `newtonsoft.json\13.0.1\lib\netstandard2.0\Newtonsoft.Json.dll` is silently replaced with `newtonsoft.json\13.0.2\lib\net6.0\Newtonsoft.Json.dll`:

### Expected Behavior

Files in the NuGet cache remain untouched.

### Actual Behavior

Files in a NuGet package are silently replaced with files from another version.

### Analysis

Both Windows and Linux systems are affected, also it doesn't matter whether hard or symbolic links are used. The problem is that `File.Copy` operation, instead of replacing the link, replaces the file that the link is linking to. To safely replace a link with a different file or link, `File.Delete` needs to be called first. Unfortunately, MSBuild calls [File.Delete](https://github.com/dotnet/msbuild/blob/a6f6699d1f70bf79db82030938d2c5e52d1e4d2e/src/Tasks/Copy.cs#L355-L360) only when the usage of hard or symbolic links is requested. When the build doesn't use hard or symbolic links, then the `File.Delete` is not called.

**Comment by rainersigwald** [comment]
Reopening since the change was backed out due to causing https://github.com/dotnet/msbuild/issues/8684.

**Comment by manfred-brands** [comment]
Hi @marcin-krystianc,

This is the first time I heard it was reverted. I didn't get any GitHub notifications, nor can I see why it was reverted. Do you know more?

**Comment by manfred-brands** [comment]
I could add an 'is same file' check in the copy code, but shouldn't msbuild not detect that itself before even calling Copy? The destination file exists and has the same timestamp.

**Comment by marcin-krystianc** [comment]
I didn't look into details yet, but the https://github.com/dotnet/msbuild/issues/8684 @Forgind explains that msbuild checks whether paths are equal before running the Copy operation, but the checks are on relative paths not on the full paths.

**Comment by manfred-brands** [comment]
I made a small test program, that emulates a copy to oneself to see how `File.Copy` behaves when this is done.

I expected one of three outcomes:
1. An exception
2. An empty destination file (as it is truncated before copying)
3. It works as the System recognizes this.

```csharp
File.Copy("Program.cs", "Program.cs", true);
```

It throws:
```
Unhandled exception. System.IO.IOException: The process cannot access the file 'C:\Users\m.brands\source\repos\CopyToSelf\Program.cs' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at Program.<Main>$(String[] args) in C:\Users\m.brands\source\repos\CopyToSelf\Program.cs:line 4
```

This would mean that in the current msbuild, this situation already throws an exception.
In the current build this exception is handled in `DoCopyWithRetries`:

```csharp

**Comment by manfred-brands** [comment]
My proposed change is:
```csharp
            // If the destination file is a hard or symbolic link, File.Copy would overwrite the source.
            // To prevent this, we need to delete the existing entry before we Copy or create a link.
            // We could try to figure out if the file is a link, but I can't think of a reason to not simply delete it always.
            if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_6) && destinationFileState.FileExists && !destinationFileState.IsReadOnly)
            {
                if (PathsAreIdentical(sourceFileState.Name, destinationFileState.Name))
                {
                    return true;
                }

                FileUtilities.DeleteNoThrow(destinationFileState.Name);
            }
```

**Comment by manfred-brands** [comment]
@Forgind The alternative I had was to replace the string.Equals on the specified paths in `DoCopyIfNecessary` with a `PathsAreIdentical`. 
Is `Path.GetFullPath` that expensive?

**Comment by danmoseley** [comment]
@JeremyKuhne GetFullPath is still significantly expensive, right? Although perhaps better than before, it will still hit the disk in some cases, right?

**Comment by manfred-brands** [comment]
It looks like the .net runtime on Windows calls into a system call GetFullPathW which has a [remark](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfullpathnamew#remarks) that it does not check if the file exists. It should only do path magic: prefix with current directory and remove unnecessary directory up and down: `A/B/../C' becomes 'A/C'.

On Linux, `GetFullPath` does a `Combine(GetCwd(), path)` and then calls `PathInternal.RemoveRelativeSegments`.

I'm surprised at the amount of Window specific (directly into OS) calls in MSBuild for the file system access.

**Comment by JeremyKuhne** [comment]
> GetFullPath is still significantly expensive, right? Although perhaps better than before, it will still hit the disk in some cases, right?

@danmoseley On Windows it typically is just a string parsing/manipulation routine. It will get the current directory and possibly environment variables (for drive relative paths), but those are not disk based. There is some sort of check regarding legacy device names `PRN`, `CON`, etc. but I don't recall precisely what it does off the top of my head (check for device availability for serial ports maybe?).

On .NET Framework (and early .NET Core versions) `Path.GetFullPath` was significantly more expensive as we would try to parse the path and check for validity up front (before passing it to the OS). In 4.6.2 I ported back some of the changes which improves things dramatically.

**Comment by Forgind** [comment]
> @Forgind The alternative I had was to replace the string.Equals on the specified paths in `DoCopyIfNecessary` with a `PathsAreIdentical`. Is `Path.GetFullPath` that expensive?
> 
> I see that is what you did in #8685.

Right; I couldn't think of another way to validate that the paths really are the same, but I moved the check a little earlier and tried to unify some of the computation to minimize the impact in most cases. The only case where I think perf regresses is if you try to create a hard/symlink, and that's minor, per JeremyKuhne's comment, unless someone is using an older runtime. I undrafted the PR; from internal discussion, it sounds like we're leaning towards taking it, so that should hopefully resolve the writing-through-symlink problem without reintroducing the copy-onto-self-means-delete problem.
